server/homebrew.api.js

```diff
@@ -89,148 +129,177 @@ const newBrew = async (req, res)=>{
+	newHomebrew.editId = nanoid(12);
+	newHomebrew.shareId = nanoid(12);
+
+	let googleId, saved;
 	if(transferToGoogle) {
-		saved = await newGoogleBrew(req.account, brew, res)
+		googleId = await newGoogleBrew(req.account, newHomebrew, res)
 			.catch((err)=>{
-				res.status(err.status || err.response.status).send(err.message || err);
+				res.status(err?.status || err?.response?.status || 500).send(err?.message || err);
 			});
+		excludeStubProps(newHomebrew);
 	} else {
-		saved = await newLocalBrew(brew)
-			.catch((err)=>{
-				res.status(500).send(err);
-			});
+		// Compress brew text to binary before saving
+		newHomebrew.textBin = zlib.deflateRawSync(newHomebrew.text);
+		// Delete the non-binary text field since it's not needed anymore
+		newHomebrew.text = undefined;
+	}
+	if(transferToGoogle && !googleId) {
+		if(!res.headersSent) {
+			res.status(500).send('Unable to save document to Google Drive');
+		}
+		return;
+	} else if(transferToGoogle) {
+		newHomebrew.googleId = googleId;
 	}
+
+	saved = await newHomebrew.save()
+		.catch((err)=>{
+			console.error(err, err.toString(), err.stack);
+			throw `Error while creating new brew, ${err.toString()}`;
+		});
 	if(!saved) return;
+	saved = saved.toObject();
+
 	return res.status(200).send(saved);
 };
 
 const updateBrew = async (req, res)=>{
-	let brew = excludePropsFromUpdate(req.body);
+	console.log(req.brew, req.body);
+	let brew = Object.assign(req.brew, excludePropsFromUpdate(req.body));
+	brew.text = mergeBrewText(brew);
 	const { transferToGoogle, transferFromGoogle } = req.query;
 
-	let saved;
 	if(brew.googleId && transferFromGoogle) {
-		beforeNewSave(req.account, brew);
-
-		saved = await newLocalBrew(brew)
+		const deleted = await deleteGoogleBrew(req.account, brew.googleId, brew.editId, res)
 			.catch((err)=>{
 				console.error(err);
-				res.status(500).send(err);
+				res.status(err?.status || err?.response?.status || 500).send(err.message || err);
 			});
-		if(!saved) return;
+		if(!deleted) {
+			if(res.headersSent) {
+				res.status(500).send('Unable to delete brew from Google');
+			}
+			return;
+		}
 
-		await deleteGoogleBrew(req.account, `${brew.googleId}${brew.editId}`, res)
-			.catch((err)=>{
-				console.error(err);
-				res.status(err.status || err.response.status).send(err.message || err);
-			});
+		brew.googleId = undefined;
 	} else if(!brew.googleId && transferToGoogle) {
-		saved = await newGoogleBrew(req.account, brew, res)
+		brew.googleId = await newGoogleBrew(req.account, excludeGoogleProps(_.clone(brew)), res)
 			.catch((err)=>{
 				console.error(err);
 				res.status(err.status || err.response.status).send(err.message || err);
 			});
-		if(!saved) return;
-
-		await deleteLocalBrew(req.account, brew.editId)
-			.catch((err)=>{
-				console.error(err);
-				res.status(err.status).send(err.message);
-			});
+		if(!brew.googleId) {
+			if(res.headersSent) {
+				res.status(500).send('Unable to save brew to Google');
```

Comment:
Same thing here. Seems like we could move all the `res.send` errors into `newGoogleBrew` to reduce the number of places we need to have duplicated error handling?

Reply:
If not a fan of passing the `res` object around, now that I think about it, I think our existing error handler middleware should be able to `send` the error as long as we just leave it uncaught or rethrow here.

So, instead of re-catching the error at each level (unless we need to add info to the error), we should be able to create the appropriate error object inside `newGoogleBrew`, throw it, and then just let the middleware catch it for final sending?